ServerCore/Helpers/AddUserHelper.cs

```diff
@@ -0,0 +1,85 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using ServerCore.DataModel;
+
+namespace ServerCore.Helpers
+{
+    /// <summary>
+    /// Includes methods that construct a list of users to be used in add-user pages.
+    /// </summary>
+    public static class AddUserHelper
+    {
+        public static async Task<List<PuzzleUserView>> GetNonAdminsForEvent(PuzzleServerContext puzzleServerContext, Event thisEvent)
+        {
+            List<int> adminUserIDs = await (from eventAdmin in puzzleServerContext.EventAdmins
+                                         where eventAdmin.Event == thisEvent
+                                         orderby eventAdmin.AdminID
+                                         select eventAdmin.AdminID).ToListAsync();
+
+            return await GetAllUsersExcept(puzzleServerContext, adminUserIDs);
+        }
+
+        public static async Task<List<PuzzleUserView>> GetNonAuthorsForEvent(PuzzleServerContext puzzleServerContext, Event thisEvent)
+        {
+            List<int> authorUserIDs = await (from eventAuthor in puzzleServerContext.EventAuthors
+                                            where eventAuthor.Event == thisEvent
+                                            orderby eventAuthor.AuthorID
+                                            select eventAuthor.AuthorID).ToListAsync();
+
+            return await GetAllUsersExcept(puzzleServerContext, authorUserIDs);
+        }
+
+        public static async Task<List<PuzzleUserView>> GetNonTeamMembersForTeam(PuzzleServerContext puzzleServerContext, Team team)
+        {
+            List<int> memberUserIDs = await (from teamMember in puzzleServerContext.TeamMembers
+                                             where teamMember.Team == team
+                                             orderby teamMember.Member.ID
+                                             select teamMember.Member.ID).ToListAsync();
+
+            return await GetAllUsersExcept(puzzleServerContext, memberUserIDs);
+        }
+
+        private static async Task<List<PuzzleUserView>> GetAllUsersExcept(PuzzleServerContext puzzleServerContext, List<int> userIDsToSkip)
+        {
+            // get all users (many)
+            List<PuzzleUserView> users = await (from user in puzzleServerContext.PuzzleUsers
```

Comment:
This looks pretty expensive since it pulls all users from all events ever. It's likely more performant to do an antijoin like this https://stackoverflow.com/a/57217382

Reply:
At this time the user may not be associated with the event at all. All of these lists currently show "all users from all events ever", minus maybe 25 at most. The number of users we are excluding from the list is very small - either the number of admins for the event, number of authors for the event, or number of players on the team. I think all the SQL overhead here is likely to be overkill since SQL has no way of knowing that removing entries from "all users ever" is the exceptional case for these queries.

Reply:
You should be able to see this in the three queries I'm _removing_, which select all users but a very small subset:

```
                Users = await (from user in _context.PuzzleUsers
                               where !((from eventAdmin in _context.EventAdmins
                                       where eventAdmin.Event == Event
                                       where eventAdmin.Admin == user
                                       select eventAdmin).Any())
                               select user).ToListAsync();

                Users = await (from user in _context.PuzzleUsers
                               where !((from eventAuthor in _context.EventAuthors
                                       where eventAuthor.Event == Event
                                       where eventAuthor.Author == user
                                       select eventAuthor).Any())
                               select user).ToListAsync();

                Users = await (from user in _context.PuzzleUsers
                            where !((from teamMember in _context.TeamMembers
                                    where teamMember.Team.Event == Event
                                    where teamMember.Member == user
                                    select teamMember).Any())
                            select new Tuple<PuzzleUser, int>(user, -1)).ToListAsync();
```

Reply:
[Correction: I guess the prior select team member one ignores all members of _all_ teams, not just the selected team, but even so that is only going to be like 10% of the user list and dropping]